packages/plugins/foundry-contracts/ChugSplashLocalExecutor.sol

```diff
@@ -0,0 +1,203 @@
+    function findMaxBatchSize(
+        BundledChugSplashActions[] actions,
+        ChugSplashManager manager,
+        uint masGasLimit
+    ) public view returns (uint) {
+        // Optimization, try to execute the entire batch at once before doing a binary search
+        if (executable(actions, manager, masGasLimit)) {
+            return actions.length;
+        }
+
+        // If the full batch isn't executavle, then do a binary search to find the largest executable batch size
+        uint min = 0;
+        uint max = actions.length;
+        while (min < max) {
+            uint mid = convert((convert(min + max) / 2).ceil());
+            BundledChugSplashActions[] memory left = actions.slice(0, mid);
+            if (executable(left, manager, masGasLimit)) {
+                min = mid;
+            } else {
+                max = mid - 1;
+            }
+        }
+        uint mid = (min + max) / 2;
+
+        // No possible size works, this is a problem and should never happen
+        if (min == 0) {
+            revert("Unable to find a batch size that does not exceed the block gas limit");
+        }
+
+        return min;
+    }
+
+    /**
+     * Helper function for executing a list of actions in batches.
+     */
+    function executeBatchActions(
+        BundledChugSplashActions[] actions,
+        ChugSplashManager manager
+    ) public {
+        // Pull he deployment state from the contract to make sure we're up to date
+        bytes32 activeDeploymentId = manager.activeDeploymentId();
+        DeploymentState storage deploymentState = manager.deployments(activeDeploymentId);
+
+        // Filter out actions that have already been executed
+        BundledChugSplashActions[] filteredActions;
+        for (uint i = 0; i < actions.length; i++) {
+            BundledChugSplashActions action = actions[i];
+            if (deploymentState.actions[action.actionIndex] == false) {
+                filteredActions.push(action);
+            }
+        }
+
+        // Exit early if there are no actions to execute
+        if (filteredActions.length == 0) {
+            return;
+        }
+
+        uint executed = 0;
+        while (executed < filteredActions.length) {
+            // Figure out the maximum number of actions that can be executed in a single batch
+            uint batchSize = findMaxBatchSize(filteredActions.slice(executed, filteredActions.length), manager, maxGasLimit);
+            BundledChugSplashActions[] memory batch = filteredActions.slice(executed, executed + batchSize);
+
+            (RawChugSplashAction[] actions, uint256[] memory _actionIndexes, bytes32[][] memory _proofs) = disassembleActions(batch);
+
+            // Execute the batch
+            // We specifically start and stop the broadcast here to make sure this is the only time transactions are actually sent
+            vm.startBroadcast();
+            manager.executeActions(actions, _actionIndexes, _proofs);
+            vm.stopBroadcast();
```

Comment:
Note the use of `vm.startBroadcast()` here, we need to do this or something like it b/c we don't want to broadcast transactions during the `findMaxBatchSize` function. I'm not sure if this is exactly how it'll need to work in practice since I haven't used Foundry much, but I know we'll need to do something like this.

Reply:
I don't think we can use `broadcast` in our solidity code because it may send transactions to the blockchain when the user doesn't intend to (e.g. on a forked network). The user must be the one to add `broadcast` to their code. This is fine because we can use `staticcall`, which doesn't broadcast anything, in `findMaxBatchSIze`.